### PR TITLE
internal: Simplify checking for supported providers and metadata function to call

### PR DESCRIPTION
**Pull request by joonas**
This duplication bothered me a bit, so I figured I'd submit a PR to get rid of it.

### CODE REVIEW

**Comment by lucab** [comment]
Perhaps a better suggestion here: introduce a `metadataFetcher(provider string) (fetchFn, error)` which returns the provider-specific fetcher or an error for unknown ones, then either react to the error or invoke the returned function. While at it, please also introduce some unit-test covering positive and negative cases.

File: internal/main.go [hunk 1]
```diff
@@ -80,14 +80,26 @@ func main() {
...
+	case "digitalocean":
+		metadataFn = digitalocean.FetchMetadata
+	case "ec2":
+		metadataFn = ec2.FetchMetadata
+	case "gce":
+		metadataFn = gce.FetchMetadata
+	case "packet":
+		metadataFn = packet.FetchMetadata
+	case "openstack-metadata":
+		metadataFn = openstackMetadata.FetchMetadata
```
Comment (lucab) [inline internal/main.go]: I'd rather see this logic splits in its own testable function, bubbling-up errors. See the general comment for details.

**Review by lucab — CHANGES_REQUESTED** [review]

**Comment by joonas** [comment]
> Perhaps a better suggestion here: introduce a metadataFetcher(provider string) (fetchFn, error) which returns the provider-specific fetcher or an error for unknown ones, then either react to the error or invoke the returned function. While at it, please also introduce some unit-test covering positive and negative cases.

I went down this path at first, but felt that it was more than I should introduce in one go. However, I'm happy to take a pass at implementing something like this if that'd be the preferred approach.

Thanks for the review!

**Comment by joonas** [comment]
@lucab - is this in line with what you were thinking?

File: internal/main.go [hunk 2]
```diff
@@ -114,6 +114,25 @@ func main() {
...
+	case "ec2":
+		return ec2.FetchMetadata, nil
+	case "gce":
+		return gce.FetchMetadata, nil
+	case "packet":
+		return packet.FetchMetadata, nil
+	case "openstack-metadata":
+		return openstackMetadata.FetchMetadata, nil
+	default:
+		return nil, errors.New("unknown provider")
```
Comment (cgonyeo) [inline internal/main.go]: This is a super tiny nit, but I prefer the pattern of doing a `var ErrUnknownProvider = ...` at the top of the file and returning it here. Makes it easier to expose known error values for tests and such.

**Review by cgonyeo — APPROVED** [review]
LGTM (Luca is current on PTO)

File: internal/main.go (hunk 2 above)
Comment (joonas) [reply internal/main.go]: Totally makes sense, just did that 👍

**Review by cgonyeo — APPROVED** [review]
LGTM, rebase this on master and I'll merge

**Comment by joonas** [comment]
Cool, should be good now 👍

File: internal/main_test.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	}{
+		{
+			desc: "supported provider",
+			name: "digitalocean",
+			err:  nil,
+		},
+		{
+			desc: "unknown provider",
+			name: "not-supported",
+			err:  errors.New("unknown provider"),
```
Comment (crawford) [inline internal/main_test.go:23]: @dgonyeo the point of that refactor was to make use of it here.
